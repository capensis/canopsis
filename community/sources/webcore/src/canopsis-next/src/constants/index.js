export * from './recurrence-rule';
export * from './common';
export * from './entity';
export * from './alarm';
export * from './broadcast-message';
export * from './date';
export * from './dynamic-info';
export * from './event-filter';
export * from './pattern';
export * from './healthcheck';
export * from './idle-rule';
export * from './junit';
export * from './modal';
export * from './pbehavior';
export * from './permission';
export * from './remediation';
export * from './scenario';
export * from './snmp';
export * from './stats';
export * from './store';
export * from './widget';
export * from './kpi';
export * from './user';
export * from './service-weather';
export * from './map';
export * from './flowchart';
export * from './parameters';
export * from './declare-ticket';
export * from './link-rule';
export * from './external-data';
export * from './tag';
export * from './counter';
export * from './calendar';
export * from './icon';
export * from './availability';
