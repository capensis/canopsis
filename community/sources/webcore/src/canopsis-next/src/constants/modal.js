export const MODALS = {
  createEvent: 'create-event',
  createAckEvent: 'create-ack-event',
  confirmAckWithTicket: 'confirm-ack-with-ticket',
  createAssociateTicketEvent: 'create-associate-ticket-event',
  createCommentEvent: 'create-comment-event',
  createChangeStateEvent: 'create-change-state-event',
  createSnoozeEvent: 'create-snooze-event',
  variablesHelp: 'variables-help',
  createEntity: 'create-entity',
  createService: 'create-service',
  entityDependenciesList: 'entity-dependencies-list',
  createEntityInfo: 'create-entity-info',
  serviceEntities: 'service-entities',
  serviceDependencies: 'service-dependencies',
  createServicePauseEvent: 'create-service-pause-event',
  editLiveReporting: 'edit-live-reporting',
  infoPopupSetting: 'info-popup-setting',
  addInfoPopup: 'add-info-popup',
  confirmation: 'confirmation',
  confirmationPhrase: 'confirmation-phrase',
  clickOutsideConfirmation: 'click-outside-confirmation',
  createWidget: 'create-widget',
  selectWidgetTemplateType: 'select-widget-template-type',
  createWidgetTemplate: 'create-widget-template',
  createFilter: 'create-filter',
  alarmsList: 'alarms-list',
  createAlarmChart: 'create-alarm-chart',
  colorPicker: 'color-picker',
  textEditor: 'text-editor',
  textEditorWithTemplate: 'text-editor-with-template',
  textFieldEditor: 'text-field-editor',
  selectView: 'select-view',
  createView: 'create-view',
  shareView: 'share-view',
  createGroup: 'create-group',
  createUser: 'create-user',
  createRole: 'create-role',
  createBroadcastMessage: 'create-broadcast-message',
  createEventFilter: 'create-event-filter',
  createMetaAlarmRule: 'create-meta-alarm-rule',
  createPattern: 'create-pattern',
  eventFilterActions: 'event-filter-actions',
  createEventFilterAction: 'create-event-filter-action',
  filtersList: 'filters-list',
  createSnmpRule: 'create-snmp-rule',
  selectViewTab: 'select-view-tab',
  createDynamicInfo: 'create-dynamic-info',
  createDynamicInfoInformation: 'create-dynamic-info-information',
  importExportViews: 'import-groups-and-views',
  dynamicInfoTemplatesList: 'dynamic-info-templates-list',
  createDynamicInfoTemplate: 'create-dynamic-info-template',
  createPlaylist: 'create-playlist',
  managePlaylistTabs: 'manage-playlist-tabs',

  pbehaviorList: 'pbehavior-list',
  pbehaviorPlanning: 'pbehavior-planning',
  pbehaviorsCalendar: 'pbehaviors-calendar',
  pbehaviorRecurrenceRule: 'pbehavior-recurrence-rule',
  createRecurrenceRule: 'create-recurrence-rule',
  selectExceptionsLists: 'select-exceptions-lists',
  pbehaviorRecurrentChangesConfirmation: 'pbehavior-recurrent-changes-confirmation',
  createPbehavior: 'create-pbehavior',
  createPbehaviorType: 'create-pbehavior-type',
  createPbehaviorReason: 'create-pbehavior-reason',
  createPbehaviorException: 'create-pbehavior-exception',
  importPbehaviorException: 'import-pbehavior-exception',

  createManualMetaAlarm: 'create-manual-meta-alarm',
  removeAlarmsFromMetaAlarm: 'remove-alarms-from-meta-alarm',
  createRemediationInstruction: 'create-remediation-instruction',
  createRemediationConfiguration: 'create-remediation-configuration',
  createRemediationJob: 'create-remediation-job',
  createRemediationInstructionsFilter: 'create-remediation-instructions-filter',
  executeRemediationInstruction: 'execute-remediation-instruction',
  executeRemediationSimpleInstruction: 'execute-remediation-simple-instruction',
  remediationPatterns: 'remediation-patterns',
  pbehaviorPatterns: 'pbehavior-patterns',
  remediationInstructionApproval: 'remediation-instruction-approval',

  imageViewer: 'image-viewer',
  imagesViewer: 'images-viewer',
  rate: 'rate',
  createScenario: 'create-scenario',
  stateSetting: 'state-setting',
  testSuite: 'test-suite',
  createIdleRule: 'create-idle-rule',
  createAlarmStatusRule: 'create-alarm-status-rule',
  healthcheckEngine: 'healthcheck-engine',
  healthcheckEnginesChainReference: 'healthcheck-engines-chain-reference',
  info: 'info',
  createKpiFilter: 'create-kpi-filter',
  createMap: 'create-map',
  createGeoMap: 'create-geo-map',
  createFlowchartMap: 'create-flowchart-map',
  createMermaidMap: 'create-mermaid-map',
  createTreeOfDependenciesMap: 'create-tree-of-dependencies-map',

  createShareToken: 'create-share-token',

  createDeclareTicketEvent: 'create-declare-ticket-event',
  createDeclareTicketRule: 'create-declare-ticket-rule',
  executeDeclareTickets: 'execute-declare-tickets',

  createLinkRule: 'create-link-rule',
};
