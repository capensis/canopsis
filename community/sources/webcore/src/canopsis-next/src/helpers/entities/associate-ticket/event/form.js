import { objectToTextPairs, textPairsToObject } from '@/helpers/text-pairs';

/**
 * @typedef {Object} AssociateTicketEvent
 * @property {string} ticket
 * @property {string} ticket_url
 * @property {string} ticket_system_name
 * @property {Object} ticket_data
 * @property {string} comment
 * @property {boolean} ticket_resources
 */

/**
 * @typedef {AssociateTicketEvent} AssociateTicketEventForm
 * @property {string} ticket_id
 * @property {string} system_name
 * @property {TextPairObject[]} mapping
 */

/**
 * Convert associate ticket event object to form compatible object
 *
 * @param {AssociateTicketEvent} [event = {}]
 * @return {AssociateTicketEventForm}
 */
export const eventToAssociateTicketForm = (event = {}) => ({
  ticket_id: event.ticket ?? '',
  ticket_url: event.ticket_url ?? '',
  ticket_resources: event.ticket_resources ?? false,
  system_name: event.ticket_system_name ?? '',
  comment: event.comment ?? '',
  mapping: objectToTextPairs(event.ticket_data),
});

/**
 * Convert form object to associate ticket API compatible object
 *
 * @param {AssociateTicketEventForm} form
 * @return {AssociateTicketEvent}
 */
export const formToAssociateTicketEvent = form => ({
  comment: form.comment,
  data: textPairsToObject(form.mapping),
  system_name: form.system_name,
  ticket: form.ticket_id,
  ticket_resources: form.ticket_resources,
  url: form.ticket_url,
});
