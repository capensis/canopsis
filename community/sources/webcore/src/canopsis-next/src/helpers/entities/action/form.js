import { omit } from 'lodash';

import { ENTITIES_STATES, ACTION_TYPES, PATTERNS_FIELDS, OLD_PATTERNS_FIELDS } from '@/constants';

import { uid } from '@/helpers/uid';
import { durationToForm } from '@/helpers/date/duration';
import { getLocaleTimezone } from '@/helpers/date/date';

import { formToPbehavior, pbehaviorToForm, pbehaviorToRequest } from '../pbehavior/form';
import { requestToForm, formToRequest } from '../shared/request/form';
import { eventToAssociateTicketForm, formToAssociateTicketEvent } from '../associate-ticket/event/form';
import {
  declareTicketRuleWebhookDeclareTicketToForm,
  formToDeclareTicketRuleWebhookDeclareTicket,
} from '../declare-ticket/rule/form';
import { filterPatternsToForm, formFilterToPatterns } from '../filter/form';

/**
 * @typedef {
 *   "snooze" |
 *   "pbehavior" |
 *   "changestate" |
 *   "webhook" |
 *   "ack" |
 *   "ackremove" |
 *   "assocticket" |
 *   "cancel"
 * } ActionType
 */

/**
 * @typedef {Object} ActionForwardAuthorParameters
 * @property {boolean} [forward_author]
 * @property {string} [author]
 */

/**
 * @typedef {ActionForwardAuthorParameters} ActionDefaultParameters
 * @property {string} output
 */

/**
 * @typedef {ActionDefaultParameters} ActionSnoozeParameters
 * @property {Duration} duration
 */

/**
 * @typedef {ActionDefaultParameters} ActionChangeStateParameters
 * @property {number} state
 */

/**
 * @typedef {ActionDefaultParameters & AssociateTicketEvent} ActionAssocTicketParameters
 */

/**
 * @typedef {Object} ActionWebhookParameters
 * @property {Request} request
 * @property {?DeclareTicketRuleWebhookDeclareTicket} [declare_ticket]
 * @property {boolean} [forward_author]
 * @property {boolean} skip_for_child
 * @property {boolean} skip_for_instruction
 * @property {string} [author]
 */

/**
 * @typedef {ActionWebhookParameters} ActionWebhookFormParameters
 * @property {RequestForm} request
 * @property {DeclareTicketRuleWebhookDeclareTicketForm} declare_ticket
 */

/**
 * @typedef {
 *   Pbehavior |
 *   ActionDefaultParameters |
 *   ActionSnoozeParameters |
 *   ActionChangeStateParameters |
 *   ActionWebhookParameters |
 *   ActionAssocTicketParameters
 * } ActionParameters
 */

/**
 * @typedef {FilterPatterns} Action
 * @property {ActionType} type
 * @property {boolean} drop_scenario_if_not_matched
 * @property {boolean} emit_trigger
 * @property {string} comment
 * @property {ActionParameters} parameters
 */

/**
 * @typedef {
 *   PbehaviorForm |
 *   ActionDefaultParameters |
 *   ActionSnoozeParameters |
 *   ActionChangeStateParameters |
 *   ActionWebhookFormParameters |
 *   ActionAssocTicketParameters
 * } ActionFormParameters
 */

/**
 * @typedef {Action} ActionForm
 * @property {FilterPatternsForm} patterns
 * @property {Object.<ActionType, ActionFormParameters>} parameters
 */

/**
 * Check action type is pbehavior
 *
 * @param {ActionType} type
 * @return {boolean}
 */
export const isPbehaviorActionType = type => type === ACTION_TYPES.pbehavior;

/**
 * Check action type is webhook
 *
 * @param {ActionType} type
 * @return {boolean}
 */
export const isWebhookActionType = type => type === ACTION_TYPES.webhook;

/**
 * Check if action type is associate ticket
 *
 * @param {ActionType} type
 * @returns {boolean}
 */
export const isAssociateTicketActionType = type => type === ACTION_TYPES.assocticket;

/**
 * Convert action parameters to form
 *
 * @param {ActionForwardAuthorParameters | {}} [parameters = {}]
 * @returns {ActionForwardAuthorParameters}
 */
const defaultActionForwardAuthorToForm = (parameters = {}) => ({
  forward_author: parameters.forward_author ?? true,
  author: parameters.author ?? '',
});

/**
 * Convert action parameters to form
 *
 * @param {ActionDefaultParameters | {}} [parameters = {}]
 * @returns {ActionDefaultParameters}
 */
const defaultActionParametersToForm = (parameters = {}) => ({
  ...defaultActionForwardAuthorToForm(parameters),
  output: parameters.output ?? '',
});

/**
 * Convert action webhook parameters to form
 *
 * @param {ActionWebhookParameters} [parameters = {}]
 * @returns {ActionWebhookFormParameters}
 */
const webhookActionParametersToForm = (parameters = {}) => ({
  ...defaultActionForwardAuthorToForm(parameters),
  declare_ticket: declareTicketRuleWebhookDeclareTicketToForm(parameters.declare_ticket),
  request: requestToForm(parameters.request),
  skip_for_child: parameters.skip_for_child ?? false,
  skip_for_instruction: parameters.skip_for_instruction ?? false,
});

/**
 * Convert action snooze parameters to form
 *
 * @param {ActionSnoozeParameters | {}} [parameters = {}]
 * @returns {ActionSnoozeParameters}
 */
const snoozeActionParametersToForm = (parameters = {}) => ({
  ...defaultActionParametersToForm(parameters),
  duration: durationToForm(parameters.duration),
});

/**
 * Convert action snooze parameters to form
 *
 * @param {ActionChangeStateParameters | {}} [parameters = {}]
 * @returns {ActionChangeStateParameters}
 */
const changeStateActionParametersToForm = (parameters = {}) => ({
  ...defaultActionParametersToForm(parameters),
  state: parameters.state ?? ENTITIES_STATES.minor,
});

/**
 * Convert action assoc ticket parameters to form
 *
 * @param {ActionAssocTicketParameters | {}} [parameters = {}]
 * @returns {ActionAssocTicketParameters}
 */
const assocTicketActionParametersToForm = (parameters = {}) => ({
  ...defaultActionParametersToForm(parameters),
  ...omit(eventToAssociateTicketForm(parameters), ['ticket_comment']),
});

/**
 * Convert action pbehavior parameters to form
 *
 * @param {Pbehavior} [parameters = {}]
 * @param {string} [timezone = getLocaleTimezone()]
 * @returns {PbehaviorForm}
 */
const pbehaviorActionParametersToForm = (parameters = {}, timezone = getLocaleTimezone()) => {
  const pbehaviorForm = pbehaviorToForm(parameters, null, timezone);

  pbehaviorForm.start_on_trigger = !!parameters.start_on_trigger;
  pbehaviorForm.duration = durationToForm(parameters.duration);

  return pbehaviorForm;
};

/**
 * Prepare parameters for all action types
 *
 * @returns {Object.<ActionType, ActionFormParameters>}
 */
const prepareDefaultActionParameters = () => ({
  [ACTION_TYPES.snooze]: snoozeActionParametersToForm(),
  [ACTION_TYPES.pbehavior]: pbehaviorActionParametersToForm(),
  [ACTION_TYPES.changeState]: changeStateActionParametersToForm(),
  [ACTION_TYPES.ack]: defaultActionParametersToForm(),
  [ACTION_TYPES.ackremove]: defaultActionParametersToForm(),
  [ACTION_TYPES.assocticket]: assocTicketActionParametersToForm(),
  [ACTION_TYPES.cancel]: defaultActionParametersToForm(),
  [ACTION_TYPES.webhook]: webhookActionParametersToForm(),
});

/**
 * Convert action parameters to form
 *
 * @param {Action} action
 * @param {string} timezone
 * @returns {Object.<ActionType, ActionParameters>}
 */
export const actionParametersToForm = (action, timezone) => {
  const parameters = prepareDefaultActionParameters();

  if (!action.type || !action.parameters) {
    return parameters;
  }

  const parametersPreparers = {
    [ACTION_TYPES.snooze]: snoozeActionParametersToForm,
    [ACTION_TYPES.webhook]: webhookActionParametersToForm,
    [ACTION_TYPES.pbehavior]: pbehaviorActionParametersToForm,
    [ACTION_TYPES.assocticket]: assocTicketActionParametersToForm,
    [ACTION_TYPES.changeState]: changeStateActionParametersToForm,
  };

  const prepareParametersToFormFunction = parametersPreparers[action.type];

  parameters[action.type] = prepareParametersToFormFunction
    ? prepareParametersToFormFunction(action.parameters, timezone)
    : defaultActionParametersToForm({ ...action.parameters });

  return parameters;
};

/**
 * Convert action to form
 *
 * @param {Action} [action = {}]
 * @param {string} [timezone = getLocaleTimezone()]
 * @returns {ActionForm}
 */
export const actionToForm = (action = {}, timezone = getLocaleTimezone()) => ({
  type: action.type ?? ACTION_TYPES.snooze,
  key: uid(),
  parameters: actionParametersToForm(action, timezone),
  drop_scenario_if_not_matched: !!action.drop_scenario_if_not_matched,
  emit_trigger: !!action.emit_trigger,
  comment: action.comment ?? '',
  patterns: filterPatternsToForm(
    action,
    [PATTERNS_FIELDS.alarm, PATTERNS_FIELDS.entity],
    [OLD_PATTERNS_FIELDS.alarm, OLD_PATTERNS_FIELDS.entity],
  ),
});

/**
 * Convert pbehavior parameters to action
 *
 * @param {ActionWebhookFormParameters | {}} [parameters = {}]
 * @return {ActionWebhookParameters}
 */
export const formToWebhookActionParameters = (parameters = {}) => ({
  declare_ticket: formToDeclareTicketRuleWebhookDeclareTicket(parameters.declare_ticket),
  request: formToRequest(parameters.request),
  skip_for_child: parameters.skip_for_child,
  skip_for_instruction: parameters.skip_for_instruction,
});

/**
 * Convert pbehavior parameters to action
 *
 * @param {PbehaviorForm | {}} [parameters = {}]
 * @param {string} [timezone = getLocaleTimezone()]
 * @return {PbehaviorRequest}
 */
export const formToPbehaviorActionParameters = (parameters = {}, timezone = getLocaleTimezone()) => {
  const pbehavior = formToPbehavior(omit(parameters, ['start_on_trigger', 'duration']), timezone);

  if (parameters.start_on_trigger) {
    pbehavior.start_on_trigger = parameters.start_on_trigger;
    pbehavior.duration = parameters.duration;
  }

  return pbehaviorToRequest(pbehavior);
};

/**
 * Convert form to action parameters
 *
 * @param {ActionForm} form
 * @param {string} [timezone]
 * @returns {ActionParameters}
 */
const formToActionParameters = (form, timezone) => {
  const {
    forward_author: forwardAuthor,
    author,
    ...parametersByCurrentType
  } = form.parameters[form.type];

  const parametersPreparers = {
    [ACTION_TYPES.webhook]: formToWebhookActionParameters,
    [ACTION_TYPES.pbehavior]: formToPbehaviorActionParameters,
    [ACTION_TYPES.assocticket]: formToAssociateTicketEvent,
  };

  const prepareParametersToAction = parametersPreparers[form.type];
  const parameters = prepareParametersToAction
    ? prepareParametersToAction(parametersByCurrentType, timezone)
    : parametersByCurrentType;

  if (!isPbehaviorActionType(form.type)) {
    parameters.forward_author = forwardAuthor;

    if (!forwardAuthor) {
      parameters.author = author;
    }
  }

  return parameters;
};

/**
 * Convert form to action
 *
 * @param {ActionForm} form
 * @param {string} [timezone]
 * @returns {Action}
 */
export const formToAction = (form, timezone) => ({
  ...omit(form, ['key', 'patterns']),
  ...formFilterToPatterns(form.patterns, [PATTERNS_FIELDS.alarm, PATTERNS_FIELDS.entity]),
  parameters: formToActionParameters(form, timezone),
});
