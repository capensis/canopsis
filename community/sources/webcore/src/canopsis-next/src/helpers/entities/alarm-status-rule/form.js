import { omit } from 'lodash';

import { PATTERNS_FIELDS, TIME_UNITS } from '@/constants';

import { durationToForm } from '@/helpers/date/duration';
import { filterPatternsToForm, formFilterToPatterns } from '@/helpers/entities/filter/form';

/**
 * @typedef {FilterPatterns} AlarmStatusRule
 * @property {string} name
 * @property {string} description
 * @property {Duration} duration
 * @property {number} priority
 * @property {number} [freq_limit]
 */

/**
 * @typedef {Object} AlarmStatusRuleForm
 * @property {string} name
 * @property {string} description
 * @property {Duration} duration
 * @property {number} priority
 * @property {FilterPatternsForm} patterns
 */

/**
 * Convert alarm status rule object to form compatible object
 *
 * @param {AlarmStatusRule | {}} [rule = {}]
 * @param {boolean} [flapping = false]
 * @return {AlarmStatusRuleForm}
 */
export const alarmStatusRuleToForm = (rule = {}, flapping = false) => {
  const form = {
    name: rule.name ?? '',
    duration: rule.duration
      ? durationToForm(rule.duration)
      : { value: 1, unit: TIME_UNITS.minute },
    priority: rule.priority,
    description: rule.description ?? '',
    patterns: filterPatternsToForm(rule, [PATTERNS_FIELDS.alarm, PATTERNS_FIELDS.entity]),
  };

  if (flapping) {
    form.freq_limit = rule.freq_limit ?? 1;
  }

  return form;
};

/**
 * Convert form compatible object to alarm status rule object
 *
 * @param {AlarmStatusRuleForm} form
 * @return {AlarmStatusRule}
 */
export const formToAlarmStatusRule = form => ({
  ...omit(form, ['patterns']),
  ...formFilterToPatterns(form.patterns),
});
