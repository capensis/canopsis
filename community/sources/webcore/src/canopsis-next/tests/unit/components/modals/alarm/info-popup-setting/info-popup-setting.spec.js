import flushPromises from 'flush-promises';
import Faker from 'faker';

import { mount, createVueInstance, shallowMount } from '@unit/utils/vue';
import { mockModals } from '@unit/utils/mock-hooks';
import { createButtonStub } from '@unit/stubs/button';
import { createFormStub } from '@unit/stubs/form';
import { createModalWrapperStub } from '@unit/stubs/modal';
import ClickOutside from '@/services/click-outside';
import { MODALS } from '@/constants';

import EditLiveReporting from '@/components/modals/alarm/info-popup-setting/info-popup-setting.vue';

const localVue = createVueInstance();

const stubs = {
  'modal-wrapper': createModalWrapperStub('modal-wrapper'),
  'date-interval-selector': true,
  'v-btn': createButtonStub('v-btn'),
  'v-form': createFormStub('v-form'),
};

const snapshotStubs = {
  'modal-wrapper': createModalWrapperStub('modal-wrapper'),
  'date-interval-selector': true,
};

const factory = (options = {}) => shallowMount(EditLiveReporting, {
  localVue,
  stubs,
  attachTo: document.body,

  parentComponent: {
    provide: {
      $clickOutside: new ClickOutside(),
    },
  },

  ...options,
});

const snapshotFactory = (options = {}) => mount(EditLiveReporting, {
  localVue,
  stubs: snapshotStubs,

  parentComponent: {
    provide: {
      $clickOutside: new ClickOutside(),
    },
  },

  ...options,
});

const selectModalActionsButtons = wrapper => wrapper.findAll('.actions button.v-btn');
const selectSubmitButton = wrapper => selectModalActionsButtons(wrapper).at(1);
const selectCancelButton = wrapper => selectModalActionsButtons(wrapper).at(0);
const selectAddPopupButton = wrapper => wrapper.find('.text button.v-btn');
const selectPopupCards = wrapper => wrapper.findAll('.text  v-card-stub');
const selectPopupCardByIndex = (wrapper, index) => selectPopupCards(wrapper).at(index);
const selectRemovePopupButtonByIndex = (wrapper, index) => selectPopupCardByIndex(wrapper, index)
  .findAll('button.v-btn').at(0);
const selectEditPopupButtonByIndex = (wrapper, index) => selectPopupCardByIndex(wrapper, index)
  .findAll('button.v-btn').at(1);

describe('info-popup-setting', () => {
  const $modals = mockModals();

  test('Form submitted after trigger submit button', async () => {
    const action = jest.fn();
    const infoPopups = [
      {
        column: Faker.datatype.string(),
        template: Faker.datatype.string(),
      },
    ];
    const config = { infoPopups, action };

    const wrapper = factory({
      propsData: {
        modal: {
          config,
        },
      },
      mocks: {
        $modals,
      },
    });

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises();

    expect(action).toBeCalledWith(infoPopups);
    expect($modals.hide).toBeCalledWith();
  });

  test('Form submitted after trigger submit button without action', async () => {
    const wrapper = factory({
      propsData: {
        modal: {
          config: {},
        },
      },
      mocks: {
        $modals,
      },
    });

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises();

    expect($modals.hide).toBeCalledWith();
  });

  test('Popup added after trigger add button', async () => {
    const action = jest.fn();
    const infoPopups = [
      {
        column: Faker.datatype.string(),
        template: Faker.datatype.string(),
      },
    ];
    const columns = [
      {
        value: Faker.datatype.string(),
      },
    ];
    const wrapper = factory({
      propsData: {
        modal: {
          config: {
            columns,
            infoPopups,
            action,
          },
        },
      },
      mocks: {
        $modals,
      },
    });

    const addPopupButton = selectAddPopupButton(wrapper);

    const newPopup = {
      column: Faker.datatype.string(),
      template: Faker.datatype.string(),
    };

    addPopupButton.trigger('click');

    expect($modals.show).toBeCalledTimes(1);
    expect($modals.show).toBeCalledWith(
      {
        name: MODALS.addInfoPopup,
        config: {
          columns,
          action: expect.any(Function),
        },
      },
    );

    const [modalArguments] = $modals.show.mock.calls[0];

    modalArguments.config.action(newPopup);

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises();

    expect(action).toBeCalledWith([
      ...infoPopups,
      newPopup,
    ]);
    expect($modals.hide).toBeCalled();
  });

  test('Popup edited after trigger edit button', async () => {
    const action = jest.fn();
    const infoPopups = [
      {
        column: Faker.datatype.string(),
        template: Faker.datatype.string(),
      },
      {
        column: Faker.datatype.string(),
        template: Faker.datatype.string(),
      },
    ];
    const columns = [
      {
        value: Faker.datatype.string(),
      },
    ];
    const wrapper = factory({
      propsData: {
        modal: {
          config: {
            columns,
            infoPopups,
            action,
          },
        },
      },
      mocks: {
        $modals,
      },
    });

    const editPopupButton = selectEditPopupButtonByIndex(wrapper, 1);

    const newPopupData = {
      column: Faker.datatype.string(),
      template: Faker.datatype.string(),
    };

    editPopupButton.trigger('click');

    expect($modals.show).toBeCalledTimes(1);
    expect($modals.show).toBeCalledWith(
      {
        name: MODALS.addInfoPopup,
        config: {
          columns,
          popup: infoPopups[1],
          action: expect.any(Function),
        },
      },
    );

    const [modalArguments] = $modals.show.mock.calls[0];

    modalArguments.config.action(newPopupData);

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises();

    expect(action).toBeCalledWith([
      infoPopups[0],
      newPopupData,
    ]);
    expect($modals.hide).toBeCalled();
  });

  test('Popup removed after trigger delete button', async () => {
    const action = jest.fn();
    const infoPopups = [
      {
        column: Faker.datatype.string(),
        template: Faker.datatype.string(),
      },
      {
        column: Faker.datatype.string(),
        template: Faker.datatype.string(),
      },
    ];
    const wrapper = factory({
      propsData: {
        modal: {
          config: {
            infoPopups,
            action,
          },
        },
      },
      mocks: {
        $modals,
      },
    });

    const removePopupButton = selectRemovePopupButtonByIndex(wrapper, 1);

    removePopupButton.trigger('click');

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises();

    expect(action).toBeCalledWith([infoPopups[0]]);
    expect($modals.hide).toBeCalled();
  });

  test('Modal hidden after trigger cancel button', async () => {
    const wrapper = factory({
      propsData: {
        modal: {
          config: {},
        },
      },
      mocks: {
        $modals,
      },
    });

    const cancelButton = selectCancelButton(wrapper);

    cancelButton.trigger('click');

    await flushPromises();

    expect($modals.hide).toBeCalled();
  });

  test('Renders `info-popup-setting` with empty modal', () => {
    const wrapper = snapshotFactory({
      propsData: {
        modal: {
          config: {},
        },
      },
      mocks: {
        $modals,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('Renders `info-popup-setting` with interval', () => {
    const wrapper = snapshotFactory({
      propsData: {
        modal: {
          config: {
            infoPopups: [
              {
                column: 'alarm.v.connector',
                template: 'Connector template',
              },
              {
                column: 'extra_details',
                template: 'Extra details template',
              },
            ],
          },
        },
      },
      mocks: {
        $modals,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
