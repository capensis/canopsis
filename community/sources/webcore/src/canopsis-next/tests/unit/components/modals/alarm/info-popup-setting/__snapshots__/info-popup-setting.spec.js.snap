// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`info-popup-setting Renders \`info-popup-setting\` with empty modal 1`] = `
<form
  class="v-form"
  novalidate="novalidate"
>
  <div
    class="modal-wrapper"
    close="close"
  >
    <div
      class="title"
    >
      <span>
        Info popup
      </span>
    </div>
     
    <div
      class="text"
    >
      <div
        class="layout justify-end"
      >
        <button
          class="v-btn v-btn--floating v-btn--icon v-btn--small theme--light secondary"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon material-icons theme--light"
            >
              add
            </i>
          </div>
        </button>
      </div>
      <div
        class="layout column"
      />
    </div>
     
    <div
      class="actions"
    >
      <button
        class="v-btn v-btn--flat theme--light"
        type="button"
      >
        <div
          class="v-btn__content"
        >
          Cancel
        </div>
      </button>
      <button
        class="primary v-btn theme--light"
        type="submit"
      >
        <div
          class="v-btn__content"
        >
          Submit
        </div>
      </button>
    </div>
  </div>
</form>
`;

exports[`info-popup-setting Renders \`info-popup-setting\` with interval 1`] = `
<form
  class="v-form"
  novalidate="novalidate"
>
  <div
    class="modal-wrapper"
    close="close"
  >
    <div
      class="title"
    >
      <span>
        Info popup
      </span>
    </div>
     
    <div
      class="text"
    >
      <div
        class="layout justify-end"
      >
        <button
          class="v-btn v-btn--floating v-btn--icon v-btn--small theme--light secondary"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon material-icons theme--light"
            >
              add
            </i>
          </div>
        </button>
      </div>
      <div
        class="layout column"
      >
        <div
          class="my-1 v-card v-card--flat v-sheet theme--light secondary white--text"
        >
          <div
            class="v-card__title"
          >
            <div
              class="layout justify-space-between"
            >
              <div>
                Column: alarm.v.connector
              </div>
              <div>
                <button
                  class="v-btn v-btn--icon v-btn--small theme--light"
                  type="button"
                >
                  <div
                    class="v-btn__content"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon material-icons theme--light error--text"
                    >
                      delete
                    </i>
                  </div>
                </button>
                <button
                  class="v-btn v-btn--icon v-btn--small theme--light"
                  type="button"
                >
                  <div
                    class="v-btn__content"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon material-icons theme--light primary--text"
                    >
                      edit
                    </i>
                  </div>
                </button>
              </div>
            </div>
          </div>
          <div
            class="v-card__text"
          >
            <p>
              Template:
            </p>
            <div
              class="v-input v-textarea v-text-field v-input--is-label-active v-input--is-dirty v-input--is-disabled theme--dark"
            >
              <div
                class="v-input__control"
              >
                <div
                  class="v-input__slot"
                >
                  <div
                    class="v-text-field__slot"
                  >
                    <textarea
                      disabled="disabled"
                      rows="5"
                    />
                  </div>
                </div>
                <div
                  class="v-text-field__details"
                >
                  <div
                    class="v-messages theme--dark"
                  >
                    <transition-group-stub
                      class="v-messages__wrapper"
                      name="message-transition"
                      tag="div"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="my-1 v-card v-card--flat v-sheet theme--light secondary white--text"
        >
          <div
            class="v-card__title"
          >
            <div
              class="layout justify-space-between"
            >
              <div>
                Column: extra_details
              </div>
              <div>
                <button
                  class="v-btn v-btn--icon v-btn--small theme--light"
                  type="button"
                >
                  <div
                    class="v-btn__content"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon material-icons theme--light error--text"
                    >
                      delete
                    </i>
                  </div>
                </button>
                <button
                  class="v-btn v-btn--icon v-btn--small theme--light"
                  type="button"
                >
                  <div
                    class="v-btn__content"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon material-icons theme--light primary--text"
                    >
                      edit
                    </i>
                  </div>
                </button>
              </div>
            </div>
          </div>
          <div
            class="v-card__text"
          >
            <p>
              Template:
            </p>
            <div
              class="v-input v-textarea v-text-field v-input--is-label-active v-input--is-dirty v-input--is-disabled theme--dark"
            >
              <div
                class="v-input__control"
              >
                <div
                  class="v-input__slot"
                >
                  <div
                    class="v-text-field__slot"
                  >
                    <textarea
                      disabled="disabled"
                      rows="5"
                    />
                  </div>
                </div>
                <div
                  class="v-text-field__details"
                >
                  <div
                    class="v-messages theme--dark"
                  >
                    <transition-group-stub
                      class="v-messages__wrapper"
                      name="message-transition"
                      tag="div"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="actions"
    >
      <button
        class="v-btn v-btn--flat theme--light"
        type="button"
      >
        <div
          class="v-btn__content"
        >
          Cancel
        </div>
      </button>
      <button
        class="primary v-btn theme--light"
        type="submit"
      >
        <div
          class="v-btn__content"
        >
          Submit
        </div>
      </button>
    </div>
  </div>
</form>
`;
